src/openvic-simulation/types/Vector.hpp

```diff
@@ -1,161 +1,42 @@
-			return *this;
-		}
-
-		constexpr friend vec2_t operator*(vec2_t const& left, vec2_t const& right) {
-			return { left.x * right.x, left.y * right.y };
-		}
-		constexpr vec2_t& operator*=(vec2_t const& right) {
-			x *= right.x;
-			y *= right.y;
-			return *this;
-		}
-		constexpr friend vec2_t operator*(vec2_t const& left, T const& right) {
-			return { left.x * right, left.y * right };
-		}
-		constexpr friend vec2_t operator*(T const& left, vec2_t const& right) {
-			return { left * right.x, left * right.y };
-		}
-		constexpr vec2_t& operator*=(T const& right) {
-			x *= right;
-			y *= right;
-			return *this;
-		}
-
-		constexpr friend vec2_t operator/(vec2_t const& left, vec2_t const& right) {
-			return { left.x / right.x, left.y / right.y };
-		}
-		constexpr vec2_t& operator/=(vec2_t const& right) {
-			x /= right.x;
-			y /= right.y;
-			return *this;
-		}
-		constexpr friend vec2_t operator/(vec2_t const& left, T const& right) {
-			return { left.x / right, left.y / right };
-		}
-		constexpr friend vec2_t operator/(T const& left, vec2_t const& right) {
-			return { left / right.x, left / right.y };
-		}
-		constexpr vec2_t& operator/=(T const& right) {
-			x /= right;
-			y /= right;
-			return *this;
-		}
-
-		constexpr friend bool operator==(vec2_t const& left, vec2_t const& right) {
-			return left.x == right.x && left.y == right.y;
-		}
-		constexpr friend bool operator!=(vec2_t const& left, vec2_t const& right) {
-			return left.x != right.x || left.y != right.y;
-		}
-
-		constexpr friend std::ostream& operator<<(std::ostream& stream, vec2_t const& value) {
-			return stream << "(" << value.x << ", " << value.y << ")";
-		}
-	};
+#include "openvic-simulation/types/fixed_point/FixedPoint.hpp"
 
-	using ivec2_t = vec2_t<int32_t>;
-	using fvec2_t = vec2_t<fixed_point_t>;
+namespace OpenVic {
 
-	static_assert(sizeof(ivec2_t) == 2 * sizeof(ivec2_t::type), "ivec2_t size does not equal the sum of its parts' sizes");
-	static_assert(sizeof(fvec2_t) == 2 * sizeof(fvec2_t::type), "fvec2_t size does not equal the sum of its parts' sizes");
+#define MAKE_VEC_ALIAS(prefix, type, size) \
+	using prefix##vec##size##_t = vec##size##_t<type>; \
+	static_assert( \
+		sizeof(prefix##vec##size##_t) == size * sizeof(type), \
+		#prefix "vec" #size "_t size does not equal the sum of its parts' sizes" \
+	);
+
+#define MAKE_VEC_ALIASES(prefix, type) \
+	MAKE_VEC_ALIAS(prefix, type, 2) \
+	MAKE_VEC_ALIAS(prefix, type, 3) \
+	MAKE_VEC_ALIAS(prefix, type, 4)
+
+	MAKE_VEC_ALIASES(i, int32_t)
+	MAKE_VEC_ALIASES(f, fixed_point_t)
+
+#undef MAKE_VEC_ALIASES
+#undef MAKE_VEC_ALIAS
+
+	// Check that everything is ordered correctly in the union
+	static_assert(offsetof(fvec4_t, x) == offsetof(fvec2_t, components[0]));
+	static_assert(offsetof(fvec4_t, y) == offsetof(fvec2_t, components[1]));
+	static_assert(offsetof(fvec4_t, z) == offsetof(fvec2_t, components[2]));
+	static_assert(offsetof(fvec4_t, w) == offsetof(fvec2_t, components[3]));
```

Comment:
Wouldn't you mean:
```suggestion
	static_assert(offsetof(fvec4_t, x) == offsetof(fvec4_t, components[0]));
	static_assert(offsetof(fvec4_t, y) == offsetof(fvec4_t, components[1]));
	static_assert(offsetof(fvec4_t, z) == offsetof(fvec4_t, components[2]));
	static_assert(offsetof(fvec4_t, w) == offsetof(fvec4_t, components[3]));
```
?